### PR TITLE
provide isError and use it in place of instanceof Error

### LINKED ISSUE #15868: provide isError and use it in place of instanceof Error
**I'm submitting a ...**

- [x] bug report
- [x] feature request
- [ ] other (Please do not submit support requests here (see above))

**Current behavior:**

angular sometimes stringifies unhandled rejections, even if they're errors.

**Expected / new behavior:**

Angular should pass them on to the correct excetion handler

**Minimal reproduction of the problem with instructions:**
Difficult to reproduce

**Angular version:** 1.x.y
1.6.3

**Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
Chrome

**Anything else:**
Provide angular.isError:
```js
function isError(V) {
  var tagName = Object.prototype.toString.call(V);
  return (
    tagName === '[object Error]' ||
    tagName === '[object Exception]' ||
    V instanceof Error
  );
}
```
MIT from https://github.com/yefremov/iserror/blob/master/index.js

and use it wherever instanceof Error was used.

### PR DESCRIPTION
**Pull request by graingert**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

feature/bug

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

Angular should pass them on to the correct excetion handler, and there's an angular.isError function now.

**Does this PR introduce a breaking change?**

No

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

re-implimented https://github.com/yefremov/iserror/blob/master/index.js to avoid license and make code faster.

Fixes #15868

### CODE REVIEW

File: src/Angular.js [hunk 1]
```diff
@@ -675,6 +676,28 @@ var isArray = Array.isArray;
 
 /**
  * @ngdoc function
+ * @name angular.isError
```
Comment (gkalpak) [inline src/Angular.js]: I don't think we want to make this public.

File: src/Angular.js [hunk 2]
```diff
@@ -675,6 +676,28 @@ var isArray = Array.isArray;
...
+ * @kind function
+ *
+ * @description
+ * Determines if a reference is an `Error`.
+ *
+ * @param {*} value Reference to check.
+ * @returns {boolean} True if `value` is an `Error`.
+ */
+function isError(value) {
+  var tag = ({}).toString.call(value);
```
Comment (gkalpak) [inline src/Angular.js]: We already have a closure-global `toString = Object.prototype.toString`. You can simply `toString.call(value)`.

**Comment by graingert** [comment]
@gkalpak what would be a good way to construct a test that boots up an iframe? And where should I put it?

```js
function testError(createError) {
    var iframe = document.createElement('iframe');
    document.body.appendChild(iframe);
    try {
        var error = createError(contentWindow);
        expect(error instanceof Error).to.be.false();
        expect(isError(error)).to.be.true();
    } finally {
        iframe.parentElement.removeChild(iframe);
    }
}

testError(function (win) {
    return new win.Error();
});

testError(function (win) {
    try {
        win.document.querySelectorAll('');
    catch(e) {
        return e;
    }
});
```

File: test/minErrSpec.js [hunk 3]
```diff
@@ -17,6 +17,7 @@ describe('minErr', function() {
   it('should return an Error factory', function() {
     var myError = testError('test', 'Oops');
     expect(myError instanceof Error).toBe(true);
+    expect(isError(myError)).toBe(true);
```
Comment (gkalpak) [inline test/minErrSpec.js]: This seems irrelevant here. Please remove.

**Review by gkalpak — COMMENTED** [review]
A couple of minor comments. Looks OK otherwise (except for the missing tests).
The tests can go in https://github.com/graingert/angular.js/blob/af80f4765e2a9e795f64103b6e15770cf6eda050/test/AngularSpec.js.

**Comment by graingert** [comment]
What's wrong with publishing it?

**Comment by gkalpak** [comment]
> What's wrong with publishing it?

We have to maintain it in a backwards compatible way. If it is not public we can change/improve/remove it without any concerns.

**Comment by graingert** [comment]
How do you not publish it?

**Comment by gkalpak** [comment]
> How do you not publish it?

Remove the `@ngdoc` anotation (and others that are not relevant/necessary: `@name`, `@module ng`, `@kind`).
(It won't be available anyway, so we don't want to have it appear on the docs either.)

File: src/ngMock/angular-mocks.js [hunk 4]
```diff
@@ -984,7 +984,7 @@ angular.mock.dump = function(object) {
     } else if (angular.isObject(object)) {
       if (angular.isFunction(object.$eval) && angular.isFunction(object.$apply)) {
         out = serializeScope(object);
-      } else if (object instanceof Error) {
+      } else if (angular.isError(object)) {
```
Comment (graingert) [inline src/ngMock/angular-mocks.js]: a publishes isError is required here.

File: test/AngularSpec.js [hunk 5]
```diff
@@ -1861,6 +1861,44 @@ describe('angular', function() {
...
+    it('should not assume objects are errors', function() {
+      expect(isError({ message: 'A fake error', stack: 'no stack here'}))
+        .toBe(false);
+    });
+
+    it('should detect simple error instances', function() {
+      expect(isError(new Error())).toBe(true);
+    });
+
+    function testError(createError) {
```
Comment (gkalpak) [inline test/AngularSpec.js]: Move this either at the top or at the bottom of the `describe` block (or even create a nested `describe` for "other-context" tests) and give it a more descriptive name (e.g. `testErrorFromDifferent context`.

Maybe having a separate `describe` block with `beforeAll`/`afterAll` blocks for creating/destorying the iframe is a good idea.

File: src/ngMock/angular-mocks.js (hunk 4 above)
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: We could make it "private API" (e.g. `angular.$$isError`), but I don't think it is worth it. This function (`angular.mock.dump`) is very rarely used. It is also very rare to have new window contexts in unit tests and it is even more rare to have them throw errors. Plus, `DOMException`s get decently stringified by `JSON.stringify` (in contrast to plain `Error`s which become `"{}"`). All that considered, I think this one occurrence can remain `object instanceof Error` :smiley:

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: test/AngularSpec.js (hunk 5 above)
Comment (graingert) [inline test/AngularSpec.js]: try { } finally { } is more fine grained in this usecase

Comment (gkalpak) [inline test/AngularSpec.js]: What do you mean?

Comment (graingert) [inline test/AngularSpec.js]: it's specific to the actual call to the testErrorFromDifferentContext function. Like the lifetime of the iframe is shorter than a test, no point leaving it hanging around until afterAll/beforeAll is called.

Comment (gkalpak) [inline test/AngularSpec.js]: I am fine either way, so feel free to leave things as they are :smiley:

Comment (graingert) [inline test/AngularSpec.js]: ok

**Comment by graingert** [comment]
@gkalpak ok, I'm pretty I've resolved anything.  Interestingly enough if seems as though Safari thinks DOMExceptions from frames are instances of Errors

File: test/AngularSpec.js [hunk 6]
```diff
@@ -1861,6 +1861,43 @@ describe('angular', function() {
...
+        var error = createError(iframe.contentWindow);
+        expect(isError(error)).toBe(true);
+      } finally {
+        iframe.parentElement.removeChild(iframe);
+      }
+    }
+
+    it('should not assume objects are errors', function() {
+      expect(isError({ message: 'A fake error', stack: 'no stack here'}))
+        .toBe(false);
```
Comment (gkalpak) [inline test/AngularSpec.js]: Nit: Unnecessarily wrapped line.

**Review by gkalpak — APPROVED** [review]
LGTM

**Comment by gkalpak** [comment]
(The commit message needs some love (as it doesn't follow [our guidelines][1]).)

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit

**Review by petebacondarwin — APPROVED** [review]

**Comment by graingert** [comment]
I'd like to add DOMError too

**Comment by gkalpak** [comment]
I am going to merge this as is, but feel free to add `DOMError` in another PR :grin:

**Comment by graingert** [comment]
<del>I'd like to add DOMError too</del>

scratch that, DOMError only has a 'name' property and never has a stack, so it shouldn't be added imhop.

**Comment by gkalpak** [comment]
Actually, before I merge it can you take care of the following:
1. Fix https://github.com/angular/angular.js/pull/15872#discussion_r113686119.
2. Fix the commit message to follow the commit message guidelines.
3. Add a comment above the function pointing to the original project.

**Comment by graingert** [comment]
> Fix the commit message to follow the commit message guidelines.

what message should I use?

**Comment by gkalpak** [comment]
Something that follows [the guidelines][1]. E.g. something like:

> fix(*): correctly detect Error instances from different contexts 
>
> 12345678901234567890123456789012345678901234567890123456789012345678901234567890
> Errors thrown from different contexts (such as an iframe or webworker) are not detected as Error
> instances and handled accordingly.
> This commit fixes it by introducing an `isError()` helper, that is able to correctly detect such
> instances.
> 
> Fixes #15868 

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit

**Comment by graingert** [comment]
@gkalpak any updates on this?

**Pull request closed without merging**